command/agent/agent.go

```diff
@@ -215,6 +216,8 @@ func (c *cmd) run(args []string) int {
 	stopCh := make(chan struct{})
 	signal.Notify(signalCh, os.Interrupt, syscall.SIGTERM, syscall.SIGHUP, syscall.SIGPIPE)
 
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
I believe you still need to cancel this in the non-signal path as well, to free resources. A `defer cancel()` would probably happen too late (which is why I refactored a bit more in #7609 so that the defer would work).

I'm guessing around line 257 would be the right place to cancel this.

Reply:
We also had another channel in use to control stopping the signal handler go routine. I went ahead and just got rid of that and used the context there instead. So calling cancel in the go routine will not only stop Agent.Start but will also shut down the signal handler. Then after the Agent.Start call I am issuing the cancel again too.